### PR TITLE
Upgrade to MozBuild 4.0

### PR DESCRIPTION
**Pull request by mrobinson**

### CODE REVIEW

File: .github/workflows/rust.yml [hunk 1]
```diff
@@ -84,22 +84,49 @@ jobs:
         components: rust-src
     - name: Install deps
       run: |
-        Start-BitsTransfer -Source https://ftp.mozilla.org/pub/mozilla/libraries/win32/MozillaBuildSetup-3.4.exe -Destination ./MozillaBuildSetup.exe
+        Start-BitsTransfer -Source https://ftp.mozilla.org/pub/mozilla/libraries/win32/MozillaBuildSetup-4.0.2.exe -Destination ./MozillaBuildSetup.exe
         .\MozillaBuildSetup.exe /S | Out-Null
-        echo 'C:\Program Files\Microsoft Visual Studio\2022\Enterprise\Tools\LLVM\bin' | Out-File -FilePath $env:GITHUB_PATH -Encoding utf8 -Append
+        iwr -useb get.scoop.sh -outfile 'install.ps1'
```
Comment (sagudev) [inline .github/workflows/rust.yml:89]: This should not be here as it was removed in https://github.com/servo/mozjs/commit/16c72d786f623f6844c315c7fe8a9fb19d713f13

File: .github/workflows/rust.yml [hunk 2]
```diff
@@ -84,22 +84,49 @@ jobs:
...
+      if: (!contains(matrix.target, 'uwp'))
+    - uses: dtolnay/rust-toolchain@master
+      if: contains(matrix.target, 'uwp')
+      with:
+        toolchain: "nightly-2023-02-02"
+        components: rust-src
     - name: Build uwp
       if: contains(matrix.target, 'uwp')
-      shell: cmd
+      env:
```
Comment (sagudev) [inline .github/workflows/rust.yml:111]: This was also refactored in https://github.com/servo/mozjs/commit/16c72d786f623f6844c315c7fe8a9fb19d713f13

**Review by sagudev — COMMENTED** [review]
BONUS: As a workaround until https://bugzilla.mozilla.org/show_bug.cgi?id=1803269 lands in tagged release I created my [own tagged release](https://github.com/sagudev/mozilla-build/releases/tag/MOZILLABUILD_4_0_42_RELEASE).

**Comment by Redfire75369** [comment]
When I tried this without switching the makefile to tabs, I'm pretty sure it had failed. Something about mozmake didn't like the formatting in the makefile.

File: .github/workflows/rust.yml [hunk 3]
```diff
@@ -132,4 +159,4 @@ jobs:
         if: success()
       - name: Mark the job as unsuccessful
         run: exit 1
-        if: "!success()"
```
Comment (sagudev) [inline .github/workflows/rust.yml:135]: No workflow runs:
```
Invalid workflow file: .github/workflows/rust.yml#L1
The workflow is not valid. .github/workflows/rust.yml: Unexpected tag '!success()'
```

File: .github/workflows/rust.yml (hunk 1 above)
Comment (mrobinson) [reply .github/workflows/rust.yml]: Thanks. I've fixed this.

File: .github/workflows/rust.yml (hunk 2 above)
Comment (mrobinson) [reply .github/workflows/rust.yml]: I've also updated this as well.

**Comment by mrobinson** [comment]
@sagudev Thanks for the review. Does this look good to you now? It seems like the CI is passing.

**Comment by mrobinson** [comment]
> When I tried this without switching the makefile from tabs, I'm pretty sure it had failed. Something about mozmake didn't like the formatting in the makefile.

Huh...that's interesting. I wonder if it was some other issue. Makefiles traditionally use tabs.

**Review by sagudev — APPROVED** [review]
LGTM

**Comment by Redfire75369** [comment]
This most likely shouldn't be merged yet. Either the next version of Mozbuild with the patch should be released or we should be using @sagudev's release.

**Comment by mrobinson** [comment]
> This most likely shouldn't be merged yet. Either the next version of Mozbuild with the patch should be released or we should be using @sagudev's release.
> 
> It's a pretty simple change at least, but I'm not sure which to go with.

Would it make sense to merge this change and then switch to @sagudev's packaged version of MozBuild in a followup PR?

**Comment by Redfire75369** [comment]
I'm mostly against merging this directly since the `start-shell.bat` and having it in the repository itself is a very hacky workaround. You should probably ask @jdm for his opinion though.

**Comment by mrobinson** [comment]
Closing this one in favor of a switch to Servo's mozillatools zip bundle.

**Pull request closed without merging**